### PR TITLE
Adding the conjugate gradient algorithm and L-BFGS to the SPIN package

### PR DESCRIPTION
**Pull request by alxvov**
**Summary**

- Implementation of conjugate gradient algorithm (Fletcher-Reeves algorithm ) and limited-memory Broyden-Fletcher-Goldfarb-Shanno for spin system.
- Adding a max_norm option to all minimisers.  max_norm - maximum force/torque on atom.

**Related Issues**

improvement of the SPIN package and of the minimizers

**Author(s)**

Aleksei Ivanov, University of Iceland, alxvov@gmail.com
Julien Tranchida, Sandia National Laboratories,  jtranch@sandia.gov 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

n/e

**Implementation Notes**
The correctness of the implementation was verified as follows. Firstly, analytical spin torques of energy were checked against numerical torques using finite difference approach. Secondly, several local minima were identified using new algorithms. These magnetic states coincide with the states obtained by dissipative spin dynamics approach. Energy of these states (including magnetic skyrmions) are in agreement with the literature.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

### CODE REVIEW

**Comment by julient31** [comment]
@akohlmey and @sjplimp , Aleksei and I implemented a new option for all minimizers. Through min_modify, users can specify whether they want to use the Euclidean norm (default and only option formerly), or a max norm option (the convergence criterion is then the max value of the 2-norms across all spins/atoms in the system). This second option is relevant as it does not depend on the size of the system, unlike the Euclidean option.

This option works for all min_styles, but hftn (we created an error message for that case).

To do that, we had to modify some of the files in src/ (min_modify.cpp, and min_styles .cpp files). Would you mind thoroughly checking this? I ran a bunch of tests, but may have missed something.

**Comment by akohlmey** [comment]
@julient31 we are past the point of accepting new features before the upcoming stable release. this will have to wait for a closer look until after the LAMMPS workshop.

**Comment by julient31** [comment]
Axel @akohlmey and Richard @rbberger, Aleksei and I have been correcting the documentation and modified the file names as we discussed. 
Axel @akohlmey, if you don't mind, I'll push the fix nve/spin modifications we discussed in a separate PR. 
Overall, I think this one is ready. Let us know if you see something you would like us to modify.

File: doc/src/min_modify.txt [hunk 1]
```diff
@@ -69,6 +71,13 @@ difference of two large values (energy before and energy after) and
 that difference may be smaller than machine epsilon even if atoms
 could move in the gradient direction to reduce forces further.
 
+The choice of a norm can be modified for the min styles {cg}, {sd}, 
+{quickmin}, {fire}, {spin}, {spin/cg} and {spin/lbfgs} using 
+the {norm} keyword.
+The default {euclidean} norm computes the 2-norm (length) of the
```
Comment (sjplimp) [inline doc/src/min_modify.txt]: Here is where to explain exactly what "max" means.

File: doc/src/min_modify.txt [hunk 2]
```diff
@@ -13,11 +13,13 @@ min_modify command :h3
...
 one or more keyword/value pairs may be listed :ulb,l
-keyword = {dmax} or {line} or {alpha_damp} or {discrete_factor}
+keyword = {dmax} or {line} or {norm} or {alpha_damp} or {discrete_factor}
   {dmax} value = max
     max = maximum distance for line search to move (distance units)
-  {line} value = {backtrack} or {quadratic} or {forcezero}
-    backtrack,quadratic,forcezero = style of linesearch to use 
+  {line} value = {backtrack} or {quadratic} or {forcezero} or {spin_cubic} or {spin_none}
+    backtrack,quadratic,forcezero,spin_cubic,spin_none = style of linesearch to use 
+  {norm} value = {euclidean} or {max}
```
Comment (sjplimp) [inline doc/src/min_modify.txt]: I suggest:
norm value = two or max
  two = Euclidean two-norm (length of 3N vector)
  max = max absolute value of any 3N vector component

I think "max" is the option you are adding?  As Aidan asked (I think)
is this the true max norm (across all components), or
are you taking the length of the 3-vector for each atom (2-norm)
and taking the max of those across all atoms?  That should be explained below.

Comment (sjplimp) [reply doc/src/min_modify.txt]: And the name of what I called the true max norm is the infinity norm, correct?
If that's what you want, best to call it that.

File: src/min_fire.cpp [hunk 3]
```diff
@@ -250,7 +250,15 @@ int MinFire::iterate(int maxiter)
     // sync across replicas if running multi-replica minimization
 
     if (update->ftol > 0.0) {
-      fdotf = fnorm_sqr();
+      if (normstyle == 1) {		// max force norm
```
Comment (sjplimp) [inline src/min_fire.cpp]: Better to use a enum value for normstyle == 1, e.g. normstyle == INFINITY or normstyle == EUCLIDEAN

Comment (sjplimp) [reply src/min_fire.cpp]: In case we add other options later.

**Comment by julient31** [comment]
@athomps , I removed the modifications in Makefile.serial. I still don't remember where this was coming from (there was a --wall).
I modified the force norms (for cg, sd, fire, and quickmin) so that they match the appropriate definition: "computes the length of the 3-vector force for each atom  (2-norm), and takes the maximum value of those across all atoms". The definitions are the same for the magnetic torques.

@sjplimp , I implemented all the changes you were asking for, but one: there is already an inf norm in Lammps (the max force component across all atoms). See min.ccp, line 803.
Instead, I created a new fnorm method, fnorm_max (see line 834 of min.cpp) for the max norm Aidan was discussing. For consistency, I let the new norm style called "max". 
From what I've seen, it seems to be fine to call this norm either "max" or "infinite".

**Comment by julient31** [comment]
@sjplimp after a brief discussion with @athomps ( ;-) ), I actually added the infinite norm too. This was actually not a lot of work, and will probably avoid some confusion. 
To summarize, the users now have three norm options:
- "two", the old one (and also the default one),
- "max", the max force norm across all atoms,
- "inf", the max force component across all atoms.
All this is documented (I provided mathematical expressions for each of the three norms) in min_modify.txt.

**Comment by akohlmey** [comment]
@julient31 @alxvov this PR has now a conflict with the master branch due to my merging of #1692. Please check it out and update accordingly. I would like to merge this PR for the next patch release.

**Comment by julient31** [comment]
@akohlmey , I did the merge and solved the conflict. Thanks a lot Axel!

**Pull request merged**